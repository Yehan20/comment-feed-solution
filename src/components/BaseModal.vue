<script setup lang="ts">

// Emits
const props = defineProps<{
  title: string,
  description: string
}>()

// Props
const emit = defineEmits<{
  '@close': [],
  '@delete': [],
}>()

</script>

<template>
  <Teleport to="#app">
    <div class="modal__wrapper">
      <div class="modal__content">
        <div class="modal__header">
          <p>{{ props.title ?? 'Your Title' }}</p>
          <button @click="emit('@close')">X</button>
        </div>
        <div class="modal__body">
          <p>
            {{ props.description ?? 'Your Description' }}
          </p>
        </div>
    
          <div class="modal__footer" >
            <button title="Keep on Writing" @click="emit('@close')">Cancel</button>
            <button title="Cancel " class="btn__discard" @click="emit('@delete')">Discard</button>
          </div>
      </div>
    </div>
  </Teleport>
</template>

<style scoped lang="scss">
@use '../assets/scss/variables.scss' as v;
@use '../assets/scss/mixins.scss' as m;

.modal__wrapper {
  position: fixed;
  background-color: rgba(0, 0, 0, 0.5);
  width: 100%;
  min-height: 100vh;
  top: 0;
  left: 0;
  bottom: 0;
  display: flex;
  justify-content: center;
  align-items: center;


  button {
    border-radius: 30px;

    &:hover {
      opacity: 0.6;
    }
  }

  p {
    margin: 0;
  }

  .modal__content {
    max-width: 450px;
    padding: 20px;
    background: v.$white;
    width: 100%;
    border-radius: 20px;
    @include m.flexConfig(space-between, nowrap, stretch, column);
    gap: 25px;

    .modal__header {
      @include m.flexConfig(space-between, nowrap, center);

      p {
        font-weight: 600;
        font-size: 1.1rem;
      }

      button {
        padding: 10px 15px;
      }

    }

    .modal__body {
      p {
        color: v.$darkGray;
        font-size: 0.9rem;
        line-height: 1.5;
      }
    }

    .modal__footer {
      @include m.flexConfig(flex-start, nowrap, center);
      gap: 10px;

      button {
        display: inline-block;
        padding: 10px 30px;

        &.btn__discard {
          background-color: v.$brand-color;
          color: v.$white;
        }

      }
    }
  }

}
</style>
