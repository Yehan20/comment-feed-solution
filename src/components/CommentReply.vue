<script setup lang="ts">

import { PxMessageReply, BiArrowUpCircle, BiArrowDownCircle } from "oh-vue-icons/icons";
import { addIcons } from 'oh-vue-icons';
import TextEditor from "./TextEditor.vue";
import { onMounted, onUpdated, ref } from "vue";
import { usecommentFeedStore } from "../store/commentFeedStore"
import { helperHasNestedMessages } from "../utils/messageHelper";
import BaseModal from "./BaseModal.vue";
import CommentVoteButtonGroup from "./CommentVoteButtonGroup.vue";


// Icom Method
addIcons(PxMessageReply, BiArrowDownCircle, BiArrowUpCircle);

// Props
const props = defineProps<{
    messageInfo: {
        id: number,
        upvoted: boolean,
        downvoted: boolean,
        parentId: number | null
        userName: string
    }
}>()


// Store
const commentFeedStore = usecommentFeedStore();


// refs
const canReplyComment = ref(false);

const hasNestedThreads = ref(helperHasNestedMessages(props.messageInfo.parentId, commentFeedStore.commentFeed, props.messageInfo.id))

const showCommentThread = ref(helperHasNestedMessages(props.messageInfo.parentId, commentFeedStore.commentFeed, props.messageInfo.id))// main rest


const showModel = ref(false);

// Local varaible

let message = 0;

onMounted(() => {
    //  console.log('mount comment',canReplyComment.value);
})


//Methods
const toggleReply = () => {
    canReplyComment.value = !canReplyComment.value
}

// Expland and Collapse Nested messages 
const handleCollapseThread = () => {
    showCommentThread.value = !showCommentThread.value
}

// Collapse the Messages in the level of the nevly added message 
const autoCollapseCommentThread = () => {

    if (showCommentThread.value) {
        showCommentThread.value = false
    }

}
const deleteCurrentMessage = () => {
    canReplyComment.value = showModel.value = false
}

const handleCancel = () => {

    // if message is empty   remove  the message if not show the model 
    (message) ? showModel.value = true :
        canReplyComment.value = !canReplyComment.value

}

// Updated Life Cycle

onUpdated(() => {
    hasNestedThreads.value = helperHasNestedMessages(props.messageInfo.parentId, commentFeedStore.commentFeed, props.messageInfo.id)
})

</script>
<template>
    <div class="comment__item__reply">

        <div class="comment__buttons__wrapper">
            <div class="bg-white rounded">

                <CommentVoteButtonGroup 
                 @@downVote="commentFeedStore.downvoteComment(props.messageInfo.id)"
                 @@upVote="commentFeedStore.upvoteComment(props.messageInfo.id)"
                 :upvoted="props.messageInfo.upvoted" 
                 :downvoted="props.messageInfo.downvoted"/>

            </div>


            <button class="btn" @click="handleCancel" v-if="canReplyComment"
                :title="canReplyComment ? 'Click to Cancel' : 'Click to Reply'">
                Cancel X
            </button>

            <button class="btn" title="Click to Reply" v-else @click="canReplyComment = true">
                Reply <v-icon scale=1 name="px-message-reply"></v-icon>
            </button>


            <button class="btn" v-if="hasNestedThreads" @click="handleCollapseThread"
                :title="showCommentThread ? 'Click to expand feed' : 'Click to hide feed'">
                {{ showCommentThread ? 'Expand +' : 'Hide -' }}


            </button>
        </div>


        <template v-if="canReplyComment">
            <TextEditor :parentId="messageInfo.id" :userName="messageInfo.userName ?? ''"
                @@autoCollapseNest="autoCollapseCommentThread" @@hideEditor="toggleReply"
                @@checkMessageLength="message = $event" :showDiscussion="false" />
        </template>



    </div>

    <template v-if="!showCommentThread">
        <slot />
    </template>

    <BaseModal title="Discard comment?"
        description="You have a comment in progress, are you sure you want to discard it? " v-if="showModel"
        @@close=" showModel = !showModel" @@delete="deleteCurrentMessage" />



</template>

<style scoped lang="scss">
@use '../assets/scss/variables.scss' as v;



.comment__item__reply {


    p {
        margin: 0;
        font-size: 13px;
        color: v.$darkGray;
        font-weight: 500;
    }

    .comment__buttons__wrapper {
        display: flex;
        gap: 10px;
        margin-bottom: 5px;
        button{
        
                text-transform: lowercase;
                border-radius: 20px;
                padding-left: 10px;
                padding-right: 10px;
   
                &.btn__upvote {
   
                   padding: 5px;
                   margin-right: 0;
               }
       
               &.btn__downvote {
                   padding: 5px;
                   margin-right: 0;
       
               }
          
               &.btn__upvote.selected {
                   color: v.$brand-color !important;
               }
       
               &.btn__downvote.selected {
                   color: v.$downvote-clr !important;
               }
       
               &:hover {
                   opacity: 0.7;
               }     
        }
      
    }

}
</style>
