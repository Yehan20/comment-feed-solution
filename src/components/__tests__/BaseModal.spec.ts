import { describe, it, expect, beforeEach, afterAll, beforeAll } from "vitest";
import { mount } from "@vue/test-utils";
import { createPinia, setActivePinia } from "pinia";
import BaseModal from "../BaseModal.vue";
import CommentReply from "../CommentReply.vue";
import { nextTick } from "vue";


describe('BaseModal', () => {

    beforeAll(() => {
        // Add a #app div to the document body for teleport to work
        const app = document.createElement('div');
        app.setAttribute('id', 'app');
        document.body.appendChild(app);
    });

    afterAll(() => {
        // Clean up by removing the #app element after tests
        const app = document.getElementById('app');
        if (app) {
            document.body.removeChild(app);
        }
    });
    
    beforeEach(() => {
        setActivePinia(createPinia());
    });


    // Mounting Test
    it("Component Should Render Correctly", () => {

        const wrapper = mount(BaseModal, {
            props: {
                title: "Are you Sure",
                description: "Your Want to Delete it"
            },
            attachTo: '#app'

        })

        expect(wrapper.html()).toMatchSnapshot()

    })


})
