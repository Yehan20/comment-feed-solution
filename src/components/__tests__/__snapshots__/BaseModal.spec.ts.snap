// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BaseModal > Component Should Render Correctly 1`] = `
"<!--teleport start-->
<!--teleport end-->"
`;
